src/history/CheckpointBuilder.cpp

```diff
@@ -0,0 +1,259 @@
+    mTxResults.reset();
+    mOpen = false;
+
+    auto maybeRename = [&](FileTransferInfo const& ft) {
+        if (fs::exists(ft.localPath_nogz()))
+        {
+            CLOG_INFO(History, "File {} already exists, skipping rename",
+                      ft.localPath_nogz());
+        }
+        else if (!fs::durableRename(
+                     ft.localPath_nogz_dirty(), ft.localPath_nogz(),
+                     getPublishHistoryDir(ft.getType(), mApp.getConfig())
+                         .string()))
+        {
+            throw std::runtime_error(
+                fmt::format("Failed to rename checkpoint file {}",
+                            ft.localPath_nogz_dirty()));
+        }
+    };
+
+    auto res = FileTransferInfo(FileType::HISTORY_FILE_TYPE_RESULTS, checkpoint,
+                                mApp.getConfig());
+    auto txs = FileTransferInfo(FileType::HISTORY_FILE_TYPE_TRANSACTIONS,
+                                checkpoint, mApp.getConfig());
+    auto ledger = FileTransferInfo(FileType::HISTORY_FILE_TYPE_LEDGER,
+                                   checkpoint, mApp.getConfig());
+    maybeRename(res);
+    maybeRename(txs);
+    maybeRename(ledger);
+}
+
+CheckpointBuilder::CheckpointBuilder(Application& app) : mApp(app)
+{
+}
+
+void
+CheckpointBuilder::appendTransactionSet(uint32_t ledgerSeq,
+                                        TxSetXDRFrameConstPtr const& txSet,
+                                        TransactionResultSet const& resultSet,
+                                        bool skipStartupCheck)
+{
+    ZoneScoped;
+    TransactionHistoryEntry txs;
+    txs.ledgerSeq = ledgerSeq;
+
+    if (txSet->isGeneralizedTxSet())
+    {
+        txs.ext.v(1);
+        txSet->toXDR(txs.ext.generalizedTxSet());
+    }
+    else
+    {
+        txSet->toXDR(txs.txSet);
+    }
+    appendTransactionSet(ledgerSeq, txs, resultSet);
+}
+
+void
+CheckpointBuilder::appendTransactionSet(uint32_t ledgerSeq,
+                                        TransactionHistoryEntry const& txSet,
+                                        TransactionResultSet const& resultSet,
+                                        bool skipStartupCheck)
+{
+    ZoneScoped;
+    if (!mStartupValidationComplete &&
+        ledgerSeq != LedgerManager::GENESIS_LEDGER_SEQ && !skipStartupCheck)
+    {
+        throw std::runtime_error("Startup validation not performed");
+    }
+    maybeOpen(ledgerSeq);
+
+    if (!resultSet.results.empty())
+    {
+        TransactionHistoryResultEntry results;
+        results.ledgerSeq = ledgerSeq;
+        results.txResultSet = resultSet;
+        mTxResults->durableWriteOne(results);
+        mTxs->durableWriteOne(txSet);
+    }
+}
+
+void
+CheckpointBuilder::appendLedgerHeader(LedgerHeader const& header,
+                                      bool skipStartupCheck)
+{
+    ZoneScoped;
+    if (!mStartupValidationComplete &&
+        header.ledgerSeq != LedgerManager::GENESIS_LEDGER_SEQ &&
+        !skipStartupCheck)
```

Comment:
Hmm I'm a bit nervous about two things here:

  - The use of `mStartupValidationComplete`. Is it possible to validate the `CheckpointBuilder`'s on-disk state on-construction, rather than having this half-done state where it's been constructed but might be invalid? When is that state useful?
  - Which leads me to: `skipStartupCheck`, which seems to be ignoring the flag! Does this enable appending to an invalid (non-recovered) state? That seems like something we should probably never do. It looks like it's called from the DB migration code, which I am assuming _usually_ has a trivially-valid state (no `CheckpointBuilder`-managed publish queue at all) but .. if we got half-way through building one and failed during migration, then crashed and restarted migration, would we then be writing a second copy of the data on the end of the half-migrated data?

Reply:
* To perform validation we need several App's managers to complete initialization and load a bit of state (latest LCL), so unfortunately we can't validate on construction if CheckpointBuilder is constructed at the same time as App. We _could_ initialize CheckpointBuilder lazily and then do validation on construction, but there are still cases where we want to skip validation (your second bullet).
* Yes, like you correctly wrote, `skipStartupCheck` is only used during the DB migration (that code will go away eventually btw, and so will skipStartupCheck) - the reason is that it doesn't make sense to do startup validation in this case, we're moving from SQL to files, so there's nothing to truncate. If a migration fails and rolls back, we simply remove all history directories, and redo the migration, deletion code is here:
https://github.com/stellar/stellar-core/pull/4446/files#diff-55b0f3e8b5f28727e3fd43b6af8569da025006279be1082bfcf0aeec44d20e08R107